### DISCUSSION TITLE
Auth0Client: Admin APIs require a client-type header

### DISCUSSION THREAD

**Discussion by edocrippaofficial** [mia-platform/community#248] [Q&A] [answered]
### Which section?

Marketplace

### Which is the version?

auth0-client:3.2.3

### Description

Hi, I tried to call the API for the user creation on auth0-client and the response was  
`client type not specified and no default client available`

In my configuration there are _n_ clients and there is no default client, and this is the intended behavior.  
The request for the users creation is done by a cronjob, so there is no `client-type` header because it is not coming from a logged user, but from a sort of admin.

It seems that the check for the header `client-type` is done on every request, thus invalidating my needs to have an admin script that generates the users.

Am I missing something or this is how it actually works?  
In my opinion the APIs that do not require the client connection (the ones that will use the auth0 users management api) should not be forced to have a client set in the headers.

### Actual Outcome

It checks the client-type header, or takes the default value on every request.

### Expected Outcome

It should not force to have a client in the 'admin' APIs

**Comment by fredmaggiowski** [comment]
Hi, can't you add a valid `client-type` header from your cronjob?

Client type is important to be able to know which app is being used to manage users and which connection should be used. At the time when the auth0-client was first implemented it was required to map requests to a specific connection; things may have changed to if this is the case feel free to propose a change to the service

**Reply by edocrippaofficial** [reply]
But when I call the `POST /user` the only meaningful connection is the one provided within the request (or the defaultCreateUserConnection config).
The same thing applies fot the `POST /change-password` for example.  
They use the `managementClient` configuration

It doesn't matter that the admin belong to a specified client. In my case the admin is not one of the clients.

**Reply by fredmaggiowski** [reply]
Can you please share more details about the configuration you are using? Also if you want lets schedule a brief meeting, my memory is pretty rusty in the matters of the auth0-client so I think speaking about the issue could speed up the support :)

**Reply by edocrippaofficial** [reply]
Yeah of course 😄 

This is my config:
```json
{
  "clients": {
    "clientA": {
      "auth0Url": "---",
      "clientId": "---",
      "clientSecret": "---",
      "redirectUrl": "---",
      "scopes": ["offline_access", "profile", "email", "website"],
      "audience": "",
      "supportedConnections": ["connectionA"],
      "defaultConnection": "connectionA"
    },
    "clientB": {
      "auth0Url": "---",
      "clientId": "---",
      "clientSecret": "---",
      "redirectUrl": "---",
      "scopes": ["offline_access", "profile", "email", "website"],
      "audience": "",
      "supportedConnections": ["connectionB"],
      "defaultConnection": "connectionB"
    },
    "clientN": {...}
  },
  "managementClient": {
    "auth0Url": "---",
    "clientId": "---",
    "clientSecret": "---",
    "supportedConnections": ["connectionA", "connectionB", "connectionN"]
  },
  "customClaimsNamespaces": [
    "https://mia-platform.eu/app_metadata",
    "https://mia-platform.eu/user_metadata"
  ]
}
```

My needs at the moment is to create an user on the connectionA for example, but without having to provide a client-type (because, effectively, I'm not the clientA).
I can resolve this by adding the header `client-type: clientA` when creating an user on the connectionA and adding the header `client-type: clientB` when creating an user on the connectionB, but this is a workaround just to be compliant.

The main point is that auth0-client requires you to be a client even if you only need do do admin stuff, in witch the context from the client (eg. clientId, clientSecret, redirectUrl) are useless.

**Reply by fredmaggiowski** [reply]
I understand and it makes sense; are you providing the connection you wish to create the user into in the request?
Because if this is feasible we could remove the requirement of the client when using management APIs.

**Reply by edocrippaofficial** [reply]
Yeah in the request for the `POST /user` you can pass the connection. If you don't pass it that way, you can set the `defaultCreateUserConnection` property of the `managementClient`, as stated [here](https://docs.mia-platform.eu/docs/runtime_suite/auth0-client/configuration#auth0-connection-integration).

The same applies for the `POST /change-password`, that accepts the connection inside the body of the request.

As stated before, I can implement a workaround at the moment but, ideally speaking, you can't take for granted that someone who uses the management APIs is a client himself.

**Reply by fredmaggiowski** [reply, marked as answer]
Yes, this makes sense; can you plan a fix on the service yourself? It should *not* be hard to do.

Otherwise please open an issue on the repo and let's talk with @albertotessarotto to plan it in a useful timeline
